File: 2024 1월 3주차/백준 1365.c
#include <stdio.h>

int binary_search(int sub[], int left, int right, int val) {
    while (left <= right) {
        int mid = left + (right - left) / 2;
        if (sub[mid] < val) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    return left;
}

int min_cut(int connections[], int N) {
    int sub[N];
    int length = 0;
    for (int i = 0; i < N; i++) {
        int val = connections[i];
        int pos = binary_search(sub, 0, length - 1, val);
        if (pos == length) {
            sub[length++] = val;
        } else {
            sub[pos] = val;
        }
    }
    return N - length;
}

int main() {
    int N;
    scanf("%d", &N);
    int connections[N];
    for (int i = 0; i < N; i++) {
        scanf("%d", &connections[i]);
    }
    printf("%d\n", min_cut(connections, N));
    return 0;
}

File: 2024 1월 3주차/백준 1365.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Main {
    public static int binarySearch(List<Integer> sub, int val) {
        int left = 0, right = sub.size() - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (sub.get(mid) < val) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return left;
    }

    public static int minCut(int[] connections) {
        List<Integer> sub = new ArrayList<>();
        for (int val : connections) {
            int pos = binarySearch(sub, val);
            if (pos == sub.size()) {
                sub.add(val);
            } else {
                sub.set(pos, val);
            }
        }
        return connections.length - sub.size();
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int N = scanner.nextInt();
        int[] connections = new int[N];
        for (int i = 0; i < N; i++) {
            connections[i] = scanner.nextInt();
        }
        System.out.println(minCut(connections));
    }
}